File: app/src/main/java/com/n30dyn4m1c/photosphere/MainActivity.java
package com.n30dyn4m1c.photosphere;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.fragment.app.Fragment;

import com.n30dyn4m1c.photosphere.camera.CaptureFragment;
import com.n30dyn4m1c.photosphere.result.ResultFragment;
import com.n30dyn4m1c.photosphere.sensor.OrientationDebugFragment;
import com.n30dyn4m1c.photosphere.storage.SphereImageStore;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Permission gate, then capture or the finished-sphere result. A DEBUG button
 * opens the orientation readout beside that gate.
 */
public class MainActivity extends AppCompatActivity
        implements CaptureFragment.Host, ResultFragment.Host {

    public static final int REQUEST_PERMISSIONS = 4101;

    private static final int STATUS_UNKNOWN = 0;
    private static final int STATUS_GRANTED = 1;
    private static final int STATUS_RATIONALE = 2;
    private static final int STATUS_PERMANENT = 3;

    private Button debugButton;
    private int permissionStatus = STATUS_UNKNOWN;
    private boolean hasRequested;
    private boolean showOrientationDebug;
    private SphereImageStore.StitchedSphere sphere;
    private final ExecutorService ioExecutor = Executors.newSingleThreadExecutor();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().setStatusBarColor(Color.TRANSPARENT);
            getWindow().setNavigationBarColor(Color.TRANSPARENT);
            getWindow().getDecorView().setSystemUiVisibility(
                    View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                            | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                            | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        }
        setContentView(R.layout.activity_main);

        debugButton = findViewById(R.id.button_orientation_debug);
        if (BuildConfig.DEBUG) {
            debugButton.setVisibility(View.VISIBLE);
            debugButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    showOrientationDebug = !showOrientationDebug;
                    showContent();
                }
            });
            ViewCompat.setOnApplyWindowInsetsListener(debugButton, new androidx.core.view.OnApplyWindowInsetsListener() {
                @Override
                public WindowInsetsCompat onApplyWindowInsets(View v, WindowInsetsCompat insets) {
                    v.setPadding(v.getPaddingLeft(), insets.getSystemWindowInsetTop() + dp(4),
                            v.getPaddingRight(), v.getPaddingBottom());
                    return insets;
                }
            });
        }

        if (hasAllPermissions(requiredPermissions())) {
            permissionStatus = STATUS_GRANTED;
        }
        showContent();
        if (permissionStatus != STATUS_GRANTED && !hasRequested) {
            hasRequested = true;
            requestPermissions(requiredPermissions(), REQUEST_PERMISSIONS);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (hasAllPermissions(requiredPermissions())) {
            permissionStatus = STATUS_GRANTED;
            if (!showOrientationDebug && !(currentFragment() instanceof ResultFragment)
                    && !(currentFragment() instanceof CaptureFragment)) {
                showContent();
            } else if (currentFragment() instanceof PermissionFragment) {
                showContent();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode != REQUEST_PERMISSIONS) {
            return;
        }
        if (hasAllPermissions(requiredPermissions())) {
            permissionStatus = STATUS_GRANTED;
        } else if (grantResults.length == 0) {
            permissionStatus = STATUS_RATIONALE;
        } else if (shouldShowAnyRationale(requiredPermissions())) {
            permissionStatus = STATUS_RATIONALE;
        } else {
            permissionStatus = STATUS_PERMANENT;
        }
        showContent();
    }

    @Override
    protected void onDestroy() {
        ioExecutor.shutdownNow();
        super.onDestroy();
    }

    @Override
    public void onSphereReady(SphereImageStore.StitchedSphere stitched) {
        sphere = stitched;
        showContent();
    }

    @Override
    public SphereImageStore.StitchedSphere getSphere() {
        return sphere;
    }

    @Override
    public void onTakeAnother() {
        discardSphere();
        showContent();
    }

    private void discardSphere() {
        SphereImageStore.StitchedSphere finished = sphere;
        sphere = null;
        if (finished == null) {
            return;
        }
        final File file = finished.file;
        ioExecutor.execute(new Runnable() {
            @Override
            public void run() {
                SphereImageStore.deleteCachedSphere(file);
            }
        });
    }

    private void showContent() {
        if (debugButton != null && BuildConfig.DEBUG) {
            debugButton.setText(showOrientationDebug
                    ? R.string.orientation_debug_close
                    : R.string.orientation_debug_open);
        }
        Fragment next;
        if (showOrientationDebug) {
            next = new OrientationDebugFragment();
        } else if (permissionStatus == STATUS_GRANTED) {
            if (sphere == null) {
                next = new CaptureFragment();
            } else {
                next = new ResultFragment();
            }
        } else {
            next = PermissionFragment.newInstance(permissionStatus);
        }
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.fragment_host, next)
                .commit();
    }

    private Fragment currentFragment() {
        return getSupportFragmentManager().findFragmentById(R.id.fragment_host);
    }

    void requestRuntimePermissions() {
        hasRequested = true;
        requestPermissions(requiredPermissions(), REQUEST_PERMISSIONS);
    }

    void openAppSettings() {
        startActivity(new Intent(
                Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                Uri.fromParts("package", getPackageName(), null)
        ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
    }

    static String[] requiredPermissions() {
        List<String> permissions = new ArrayList<String>();
        permissions.add(Manifest.permission.CAMERA);
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
            permissions.add(Manifest.permission.WRITE_EXTERNAL_STORAGE);
        }
        return permissions.toArray(new String[0]);
    }

    private boolean hasAllPermissions(String[] permissions) {
        for (int i = 0; i < permissions.length; i++) {
            if (ContextCompat.checkSelfPermission(this, permissions[i])
                    != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    private boolean shouldShowAnyRationale(String[] permissions) {
        for (int i = 0; i < permissions.length; i++) {
            if (shouldShowRequestPermissionRationale(permissions[i])) {
                return true;
            }
        }
        return false;
    }

    private int dp(int value) {
        return Math.round(value * getResources().getDisplayMetrics().density);
    }

    /**
     * Front-door permission screen shown until the camera grant lands.
     */
    public static class PermissionFragment extends Fragment {

        private static final String ARG_STATUS = "status";

        static PermissionFragment newInstance(int status) {
            PermissionFragment fragment = new PermissionFragment();
            Bundle args = new Bundle();
            args.putInt(ARG_STATUS, status);
            fragment.setArguments(args);
            return fragment;
        }

        public PermissionFragment() {
            super(R.layout.fragment_permission);
        }

        @Override
        public void onViewCreated(View view, Bundle savedInstanceState) {
            super.onViewCreated(view, savedInstanceState);
            int status = getArguments() != null ? getArguments().getInt(ARG_STATUS) : STATUS_UNKNOWN;
            TextView message = view.findViewById(R.id.permission_message);
            Button action = view.findViewById(R.id.permission_action);
            final MainActivity activity = (MainActivity) requireActivity();
            if (status == STATUS_PERMANENT) {
                message.setText(R.string.permission_camera_denied_permanently);
                action.setText(R.string.permission_open_settings);
                action.setVisibility(View.VISIBLE);
                action.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        activity.openAppSettings();
                    }
                });
            } else if (status == STATUS_RATIONALE) {
                message.setText(R.string.permission_camera_rationale);
                action.setText(R.string.permission_grant);
                action.setVisibility(View.VISIBLE);
                action.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        activity.requestRuntimePermissions();
                    }
                });
            } else {
                message.setText(R.string.permission_camera_rationale);
                action.setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/com/n30dyn4m1c/photosphere/MainActivity.kt
package com.n30dyn4m1c.photosphere

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import androidx.activity.ComponentActivity
import androidx.activity.SystemBarStyle
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.systemBarsPadding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.PhotoCamera
import androidx.compose.material3.Button
import androidx.compose.material3.FilledTonalButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import androidx.lifecycle.compose.LifecycleResumeEffect
import com.n30dyn4m1c.photosphere.camera.PhotoSphereCameraScreen
import com.n30dyn4m1c.photosphere.result.PanoramaResultScreen
import com.n30dyn4m1c.photosphere.sensor.OrientationDebugScreen
import com.n30dyn4m1c.photosphere.storage.SphereImageStore
import com.n30dyn4m1c.photosphere.storage.SphereImageStore.StitchedSphere
import com.n30dyn4m1c.photosphere.ui.theme.PhotoSphereTheme
import com.n30dyn4m1c.photosphere.ui.theme.PillShape
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.launch

class MainActivity : ComponentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Both bars stay transparent over the viewfinder, and both are forced to
        // *light* icons. The default follows the system's day/night setting,
        // which on a phone in light mode paints dark status-bar icons — over a
        // black viewfinder, or a night scene, that is a clock nobody can read.
        // The app's own scheme is dark in every configuration (see
        // PhotoSphereTheme), so the bars are pinned to match it rather than the
        // system.
        enableEdgeToEdge(
            statusBarStyle = SystemBarStyle.dark(Color.TRANSPARENT),
            navigationBarStyle = SystemBarStyle.dark(Color.TRANSPARENT),
        )
        setContent {
            PhotoSphereTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background,
                ) {
                    PhotoSphereApp()
                }
            }
        }
    }

    companion object {
        /**
         * Runtime (dangerous) permissions this app needs.
         *
         * `HIGH_SAMPLING_RATE_SENSORS` is deliberately absent: it is a *normal*
         * permission granted at install time, so requesting it at runtime always
         * fails. Declaring it in the manifest is all that is required.
         *
         * Storage is only requested on API 28 and below. From API 29 the app
         * writes through MediaStore, which needs no permission for its own media.
         */
        val REQUIRED_PERMISSIONS: List<String> = buildList {
            add(Manifest.permission.CAMERA)
            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
                add(Manifest.permission.WRITE_EXTERNAL_STORAGE)
            }
        }
    }
}

/**
 * The whole app: capture, then what to do with what came out.
 *
 * There are only two destinations, so this is a `when` over one piece of state
 * rather than a navigation graph. The sphere held here is the app's single
 * source of truth for "there is a finished photo waiting" — non-null puts the
 * result screen up, and clearing it deletes the cached JPEG and returns to
 * capture with a fresh buffer.
 *
 * The camera screen is behind the permission gate; the result screen is not
 * separately gated, since the only way to reach it is through a capture that
 * already passed.
 */
@Composable
private fun PhotoSphereApp(modifier: Modifier = Modifier) {
    val scope = rememberCoroutineScope()

    // The orientation readout needs no permissions, so it sits beside the
    // permission gate rather than behind it.
    var showOrientationDebug by rememberSaveable { mutableStateOf(false) }

    // Deliberately not `rememberSaveable`: on a configuration change the file is
    // still in the cache, but a File is not something to smuggle through a
    // Bundle, and re-showing a stale result would be worse than starting over.
    var sphere by remember { mutableStateOf<StitchedSphere?>(null) }

    /** Throws away the finished sphere and goes back to capture. */
    fun discardSphere() {
        val finished = sphere ?: return
        sphere = null
        // NonCancellable: this runs as the result screen is being torn down, and
        // a several-megabyte JPEG left in the cache is exactly what the next
        // stitch would have to clean up.
        scope.launch(Dispatchers.IO + NonCancellable) {
            SphereImageStore.deleteCachedSphere(finished.file)
        }
    }

    Box(modifier = modifier.fillMaxSize()) {
        when {
            showOrientationDebug -> OrientationDebugScreen()

            else -> RequirePermissions(
                permissions = MainActivity.REQUIRED_PERMISSIONS,
            ) {
                val finished = sphere
                if (finished == null) {
                    PhotoSphereCameraScreen(onSphereReady = { sphere = it })
                } else {
                    PanoramaResultScreen(
                        sphere = finished,
                        onTakeAnother = { discardSphere() },
                    )
                }
            }
        }

        // Debug-only entry point. BuildConfig.DEBUG is a compile time constant,
        // so R8 drops this from release builds.
        if (BuildConfig.DEBUG) {
            FilledTonalButton(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .systemBarsPadding()
                    .padding(12.dp),
                onClick = { showOrientationDebug = !showOrientationDebug },
            ) {
                Text(
                    stringResource(
                        if (showOrientationDebug) {
                            R.string.orientation_debug_close
                        } else {
                            R.string.orientation_debug_open
                        }
                    )
                )
            }
        }
    }
}

/** Where the user currently stands with a set of runtime permissions. */
private enum class PermissionStatus {
    /** Not asked yet, or the system will show the dialog on the next request. */
    Unknown,

    /** Every permission in the set is granted. */
    Granted,

    /** Denied once; the system will still show the dialog, so explain why first. */
    ShowRationale,

    /** Denied with "don't ask again" (or blocked by policy) — only Settings helps. */
    PermanentlyDenied,
}

/**
 * Gates [content] behind the given runtime permissions.
 *
 * The request is fired once when the gate first appears. If the user denies it,
 * a rationale with a retry button is shown; if the OS will no longer surface the
 * dialog, the button deep-links into the app's settings page instead. Returning
 * from Settings re-checks the grant, so the UI recovers without a restart.
 */
@Composable
private fun RequirePermissions(
    permissions: List<String>,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    val context = LocalContext.current
    val activity = context.findActivity()

    var status by remember {
        mutableStateOf(
            if (context.hasAllPermissions(permissions)) {
                PermissionStatus.Granted
            } else {
                PermissionStatus.Unknown
            }
        )
    }
    var hasRequested by remember { mutableStateOf(false) }

    val launcher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { results ->
        status = when {
            // Asked of the system, not inferred from `results`. When the dialog
            // is dismissed without an answer — a swipe away, a phone call taking
            // the foreground, the screen locking — the contract delivers an
            // *empty* map, and `emptyMap().values.all { }` is vacuously true. Read
            // straight from the map, that cancellation would have counted as a
            // grant and dropped the user into a camera screen with no camera
            // permission, where the bind fails and the viewfinder is simply black.
            context.hasAllPermissions(permissions) -> PermissionStatus.Granted

            // A real denial the system will still show a dialog for, or that same
            // cancellation. Both leave the user somewhere they can retry from,
            // which is what matters: `shouldShowRationale` reads false after a
            // cancel, so trusting it here would send someone who never answered
            // the dialog off to the Settings app to fix a setting they had not
            // touched. Only an explicit "deny" that the system will no longer
            // surface a dialog for earns that trip.
            results.isEmpty() -> PermissionStatus.ShowRationale

            activity != null && permissions.any(activity::shouldShowRationale) ->
                PermissionStatus.ShowRationale

            else -> PermissionStatus.PermanentlyDenied
        }
    }

    // Ask once, automatically, the first time the gate is shown.
    LaunchedEffect(permissions) {
        if (status != PermissionStatus.Granted && !hasRequested) {
            hasRequested = true
            launcher.launch(permissions.toTypedArray())
        }
    }

    // Re-check on resume: the user may have granted the permission in Settings.
    LifecycleResumeEffect(permissions) {
        if (context.hasAllPermissions(permissions)) {
            status = PermissionStatus.Granted
        }
        onPauseOrDispose { }
    }

    when (status) {
        PermissionStatus.Granted -> content()

        PermissionStatus.Unknown -> PermissionMessage(
            modifier = modifier,
            message = stringResource(R.string.permission_camera_rationale),
            actionLabel = null,
            onAction = null,
        )

        PermissionStatus.ShowRationale -> PermissionMessage(
            modifier = modifier,
            message = stringResource(R.string.permission_camera_rationale),
            actionLabel = stringResource(R.string.permission_grant),
            onAction = { launcher.launch(permissions.toTypedArray()) },
        )

        PermissionStatus.PermanentlyDenied -> PermissionMessage(
            modifier = modifier,
            message = stringResource(R.string.permission_camera_denied_permanently),
            actionLabel = stringResource(R.string.permission_open_settings),
            onAction = { context.openAppSettings() },
        )
    }
}

/**
 * The permission gate's own screen.
 *
 * This is the app's front door — for a first-time user it is the whole app
 * until they say yes — so it is composed rather than dumped: the lens glyph
 * gives the request a subject, the title carries the ask, and the body explains
 * why a camera app that never uploads anything still needs the camera. The
 * action, when there is one, is full-width at the bottom where a thumb is.
 */
@Composable
private fun PermissionMessage(
    message: String,
    actionLabel: String?,
    onAction: (() -> Unit)?,
    modifier: Modifier = Modifier,
) {
    Scaffold(
        modifier = modifier.fillMaxSize(),
        containerColor = MaterialTheme.colorScheme.background,
    ) { insets ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(insets)
                .padding(horizontal = 32.dp, vertical = 24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
        ) {
            // A ring around the lens glyph, echoing the capture reticle the user
            // is about to spend the next few minutes aiming.
            Box(
                modifier = Modifier
                    .size(88.dp)
                    .clip(CircleShape)
                    .background(MaterialTheme.colorScheme.primary.copy(alpha = 0.12f)),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = Icons.Outlined.PhotoCamera,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.size(38.dp),
                )
            }

            Text(
                modifier = Modifier.padding(top = 28.dp),
                text = stringResource(R.string.permission_camera_title),
                style = MaterialTheme.typography.headlineSmall,
                color = MaterialTheme.colorScheme.onBackground,
                textAlign = TextAlign.Center,
            )
            Text(
                modifier = Modifier.padding(top = 12.dp),
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
            )
            if (actionLabel != null && onAction != null) {
                Button(
                    modifier = Modifier
                        .padding(top = 32.dp)
                        .fillMaxWidth()
                        .height(54.dp),
                    shape = PillShape,
                    onClick = onAction,
                ) {
                    Text(
                        text = actionLabel,
                        style = MaterialTheme.typography.titleMedium,
                    )
                }
            }
        }
    }
}

private fun Context.hasAllPermissions(permissions: List<String>): Boolean =
    permissions.all {
        ContextCompat.checkSelfPermission(this, it) == PackageManager.PERMISSION_GRANTED
    }

private fun Activity.shouldShowRationale(permission: String): Boolean =
    shouldShowRequestPermissionRationale(permission)

private fun Context.findActivity(): Activity? {
    var current: Context? = this
    while (current is android.content.ContextWrapper) {
        if (current is Activity) return current
        current = current.baseContext
    }
    return null
}

private fun Context.openAppSettings() {
    startActivity(
        Intent(
            Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
            Uri.fromParts("package", packageName, null),
        ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
    )
}
